File: profiler/app/backend/utils/validation.py
"""
WebSocket message validation using Pydantic.

This module provides schemas for validating WebSocket messages
to ensure consistent contracts between frontend and backend.
"""

from enum import Enum
from typing import Dict, Any, Optional, List, Union
from pydantic import BaseModel, Field, validator, AnyUrl

class ProfileStatus(str, Enum):
    IDLE = "idle"
    PROCESSING = "processing"
    COMPLETED = "completed"
    ERROR = "error"

class ProfileState(BaseModel):
    """Profile state model"""
    userId: str
    status: ProfileStatus
    progress: float = Field(..., ge=0, le=100)
    data: Optional[Dict[str, Any]] = None

class WebSocketMessage(BaseModel):
    """Base WebSocket message model"""
    type: str
    data: Optional[Any] = None
    error: Optional[str] = None
    timestamp: Optional[str] = None

class StateUpdateMessage(WebSocketMessage):
    """State update message model"""
    type: str = "state_update"
    data: ProfileState

class ConnectedMessage(WebSocketMessage):
    """Connected message model"""
    type: str = "connected"
    session_id: str
    message: str

class ErrorMessage(WebSocketMessage):
    """Error message model"""
    type: str = "error"
    error: str

class DocumentAnalysisData(BaseModel):
    """Document analysis data model"""
    documentId: str
    content: Optional[str] = None
    url: Optional[AnyUrl] = None

class DocumentAnalysisMessage(WebSocketMessage):
    """Document analysis message model"""
    type: str = "analyze_document"
    data: DocumentAnalysisData

class QuestionData(BaseModel):
    """Question data model"""
    question: str
    context: Optional[str] = None

class QuestionMessage(WebSocketMessage):
    """Question message model"""
    type: str = "ask_question"
    data: QuestionData

class RecommendationData(BaseModel):
    """Recommendation data model"""
    profile_id: Optional[str] = None
    count: Optional[int] = Field(None, gt=0)

class RecommendationMessage(WebSocketMessage):
    """Recommendation message model"""
    type: str = "get_recommendations"
    data: RecommendationData

def validate_websocket_message(message: Dict[str, Any]) -> WebSocketMessage:
    """
    Validate a WebSocket message based on its type.
    
    Args:
        message: Dictionary containing the message
        
    Returns:
        Validated WebSocketMessage instance
        
    Raises:
        ValueError: If the message is invalid
    """
    if not isinstance(message, dict) or 'type' not in message:
        raise ValueError("Invalid message format, missing 'type' field")
    
    message_type = message.get("type")
    
    # Map message types to their corresponding models
    type_to_model = {
        "state_update": StateUpdateMessage,
        "connected": ConnectedMessage,
        "error": ErrorMessage,
        "analyze_document": DocumentAnalysisMessage,
        "ask_question": QuestionMessage,
        "get_recommendations": RecommendationMessage
    }
    
    # Get the appropriate model for validation
    model = type_to_model.get(message_type, WebSocketMessage)
    
    try:
        return model(**message)
    except Exception as e:
        raise ValueError(f"Invalid message format for type '{message_type}': {str(e)}") 

File: profiler/app/ui/src/services/validation.ts
import { z } from 'zod';
import type { WebSocketMessage } from '../../../shared/types';

// Profile state schema
export const profileStateSchema = z.object({
  userId: z.string(),
  status: z.enum(['idle', 'processing', 'completed', 'error']),
  progress: z.number().min(0).max(100),
  data: z.record(z.any()).optional()
});

// Base WebSocket message schema
export const webSocketMessageSchema = z.object({
  type: z.string(),
  data: z.any().optional(),
  error: z.string().optional(),
  timestamp: z.string().optional()
});

// State update message schema
export const stateUpdateMessageSchema = webSocketMessageSchema.extend({
  type: z.literal('state_update'),
  data: profileStateSchema
});

// Connected message schema
export const connectedMessageSchema = webSocketMessageSchema.extend({
  type: z.literal('connected'),
  session_id: z.string(),
  message: z.string()
});

// Error message schema
export const errorMessageSchema = webSocketMessageSchema.extend({
  type: z.literal('error'),
  error: z.string()
});

// Document analysis message schema
export const documentAnalysisMessageSchema = webSocketMessageSchema.extend({
  type: z.literal('analyze_document'),
  data: z.object({
    documentId: z.string(),
    content: z.string().optional(),
    url: z.string().url().optional()
  })
});

// Question message schema
export const questionMessageSchema = webSocketMessageSchema.extend({
  type: z.literal('ask_question'),
  data: z.object({
    question: z.string(),
    context: z.string().optional()
  })
});

// Recommendation message schema
export const recommendationMessageSchema = webSocketMessageSchema.extend({
  type: z.literal('get_recommendations'),
  data: z.object({
    profile_id: z.string().optional(),
    count: z.number().positive().optional()
  })
});

// Map of message types to their schemas
export const messageSchemas: Record<string, z.Schema> = {
  'state_update': stateUpdateMessageSchema,
  'connected': connectedMessageSchema,
  'error': errorMessageSchema,
  'analyze_document': documentAnalysisMessageSchema,
  'ask_question': questionMessageSchema,
  'get_recommendations': recommendationMessageSchema
};

/**
 * Validate a message against its schema
 * @param message The message to validate
 * @param schema The schema to validate against
 * @returns The validated message
 * @throws If validation fails
 */
export const validateMessage = <T>(message: unknown, schema: z.Schema<T>): T => {
  return schema.parse(message);
};

/**
 * Validate a message based on its type
 * @param message The message to validate
 * @returns The validated message
 * @throws If validation fails
 */
export const validateMessageByType = (message: WebSocketMessage): WebSocketMessage => {
  const schema = messageSchemas[message.type];
  if (!schema) {
    throw new Error(`No schema found for message type: ${message.type}`);
  }
  return validateMessage(message, schema);
}; 